app/src/main/proto/vss.proto

```diff
@@ -0,0 +1,169 @@
+  /*
+    Fetched value and version along with corresponding key in request.
+   */
+  KeyValue value = 2;
+}
+
+
+message PutObjectRequest {
+
+  string storeId = 1;
+  /*
+    If present, write will only succeed if current DB globalVersion against a storeId is
+    same as in request.
+    Client is expected to store(client-side) global version against storeId.
+    And PutObjectRequest should contain their client-side value of globalVersion.
+
+    For first write of store, global version should be '0'. If write succeeds, clients must increment
+    their global version(client-side) by 1. This updated global version should be used in subsequent
+    PutObjectRequests for the store.
+
+    Requests with conflicting version will fail with [`ConflictException`] as ErrorCode.
+   */
+  optional int64 globalVersion = 2;
+
+  /*
+    Each key is supplied with its corresponding value and version.
+    Clients can choose to encrypt keys client-side in order to obfuscate user's usage patterns.
+    If write is successful, previous value of key will be overwritten.
+
+    Multiple items in transactionItems of single PutObjectRequest are written in
+    a database-transaction in all-or-nothing fashion.
+    Items in single PutObjectRequest should have distinct keys.
+
+    Write will only succeed if current DB version against key is same as in request.
+    Client is expected to store version against every key.
+    When initiating a PutObjectRequest, request should contain their client-side version for that key.
+
+    For first write of key, version should be '0'. If write succeeds, clients must increment
+    their corresponding key versions(client-side) by 1. These updated key versions should be used in
+    subsequent PutObjectRequests for the keys.
+
+    Requests with conflicting version will fail with [`ConflictException`] as ErrorCode.
+
+    Considerations for transactions:
+    Transaction writes of multiple items have a performance overhead, hence it is recommended to use
+    them only if required by client application to ensure logic/code correctness.
+    That is, transactionItems are not a substitute for batch-write of multiple unrelated items.
+    When write of multiple unrelated items is desired, it is recommended to use separate
+    PutObjectRequests.
+
+    Consistency Guarantee:
+    All PutObjectRequests are strongly consistent i.e. they provide read-after-write and
+    read-after-update consistency guarantees.
+   */
+  repeated KeyValue transactionItems = 3;
+}
+
+message PutObjectResponse {
+}
+
+
+/*
+  When HttpStatusCode is not ok(200), response `content` contains serialized ErrorResponse with
+  relevant ErrorCode and errorCode's integer value as HttpStatusCode.
+ */
+message ErrorResponse {
+  /*
+    The error code is a string that uniquely identifies an error condition.
+    It is meant to be read and understood programmatically  by code that detects/handles errors by
+    type.
+   */
+  ErrorCode errorCode = 1;
```

Comment:
Is this duplicative of the the HTTP status code? Or can they diverge?

Reply:
Rather, my question was really why include it here if it is the same as the HTTP status code?

Reply:
ErrorCode if not being used anywhere in model, could be difficult for client to understand how and where it is being used.
Also right now, model is agnostic of http protocol and status code.

Reply:
That makes sense, i will remove this association from ErrorCode enum's value. 
Since http status code to errorCode could be one to many in future.